### PR TITLE
Add Draft of swagger spec for Authoritative Server HTTP API

### PR DESCRIPTION
**Pull request by ffledgling**
### Short description
- Defines most object types
- Documents GET requests for /servers and /zones

Issue #5854 

### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] included documentation

### CODE REVIEW

File: docs/http-api/swagger/authoritative-api-swagger.yaml [hunk 1]
```diff
@@ -0,0 +1,996 @@
...
+    get:
+      summary: List all servers
+      operationId: listServers
+      tags:
+        - servers
+      responses:
+        '200':
+          description: An array of servers
+          schema:
+            #$ref: '#/definitions/Servers'
```
Comment (ffledgling) [inline docs/http-api/swagger/authoritative-api-swagger.yaml]: Dead comment

File: docs/http-api/swagger/authoritative-api-swagger.yaml [hunk 2]
```diff
@@ -0,0 +1,996 @@
...
+        - name: server_id
+          in: path
+          required: true
+          description: The id of the server to retrieve
+          type: string
+      responses:
+        '200':
+          description: An array of Zones
+          schema:
+            #$ref: '#/definitions/Zones'
```
Comment (ffledgling) [inline docs/http-api/swagger/authoritative-api-swagger.yaml]: Dead comment

File: docs/http-api/swagger/authoritative-api-swagger.yaml [hunk 3]
```diff
@@ -0,0 +1,996 @@
...
+          description: The id of the server to retrieve
+          type: string
+        - name: rrsets
+          in: query
+          description: "“true” (default) or “false”, whether to include the “rrsets” in the response Zone object."
+          type: boolean
+          default: true
+      responses:
+        '201':
+          description: An array of Zones
```
Comment (ffledgling) [inline docs/http-api/swagger/authoritative-api-swagger.yaml]: Incorrect description.

File: docs/http-api/swagger/authoritative-api-swagger.yaml [hunk 4]
```diff
@@ -0,0 +1,996 @@
...
+          required: true
+          description: The id of the zone to retrieve
+      responses:
+        '200':
+          description: OK
+
+  '/servers/{server_id}/zones/{zone_id}/export':
+    get:
+      summary: "Returns the zone in AXFR format."
+      description: ""
```
Comment (ffledgling) [inline docs/http-api/swagger/authoritative-api-swagger.yaml]: Unnecessary.

File: docs/http-api/swagger/authoritative-api-swagger.yaml [hunk 5]
```diff
@@ -0,0 +1,996 @@
...
+      responses:
+        '200':
+          description: OK
+          schema:
+            type: string
+
+  '/servers/{server_id}/zones/{zone_id}/check':
+    get:
+      summary: "Verify zone contents/configuration."
+      description: ""
```
Comment (ffledgling) [inline docs/http-api/swagger/authoritative-api-swagger.yaml]: " " "

File: docs/http-api/swagger/authoritative-api-swagger.yaml [hunk 6]
```diff
@@ -0,0 +1,996 @@
...
+          description: The id of the zone to retrieve
+      responses:
+        '200':
+          description: OK
+          schema:
+            type: string
+
+  '/servers/{server_id}/zones/{zone_id}/rectify':
+    put:
+      summary: "Rectify the zone data. This does not take into account the API-RECTIFY metadata."
```
Comment (ffledgling) [inline docs/http-api/swagger/authoritative-api-swagger.yaml]: Move second sentence from summary into description.

File: docs/http-api/swagger/authoritative-api-swagger.yaml [hunk 7]
```diff
@@ -0,0 +1,996 @@
...
+          description: The id of the zone to retrieve
+      responses:
+        '200':
+          description: OK
+          schema:
+            type: string
+
+  '/servers/{server_id}/config':
+    get:
+      summary: 'Returns all ConfigSetting for a single server'
```
Comment (ffledgling) [inline docs/http-api/swagger/authoritative-api-swagger.yaml]: Settings, plural.

File: docs/http-api/swagger/authoritative-api-swagger.yaml [hunk 8]
```diff
@@ -0,0 +1,996 @@
...
+      responses:
+        '200':
+          description: List of config values
+          schema:
+            $ref: '#/definitions/ConfigSetting'
+
+  '/servers/{server_id}/statistics':
+    get:
+      summary: 'Query PowerDNS internal statistics.'
+      description: 'Query PowerDNS internal statistics. Returns a list of StatisticItem elements.'
```
Comment (ffledgling) [inline docs/http-api/swagger/authoritative-api-swagger.yaml]: Duplication.

File: docs/http-api/swagger/authoritative-api-swagger.yaml [hunk 9]
```diff
@@ -0,0 +1,996 @@
...
+          description: 'The string to search for'
+          type: string
+        - name: max
+          in: query
+          required: true
+          description: 'Maximum number of entries to return'
+          type: integer
+      responses:
+        '200':
+          description: Returns a single JSON object with a single array of strings.
```
Comment (ffledgling) [inline docs/http-api/swagger/authoritative-api-swagger.yaml]: Description is incorrect.

File: docs/http-api/swagger/authoritative-api-swagger.yaml [hunk 10]
```diff
@@ -0,0 +1,996 @@
...
+          schema:
+            $ref: '#/definitions/SearchResults'
+
+  # TODO: Fix metadata post and return types since
```
Comment (ffledgling) [inline docs/http-api/swagger/authoritative-api-swagger.yaml]: Improve comment.

**Comment by pieterlexis** [comment]
Note: interactive docs [here](https://app.swaggerhub.com/apis/ffledgling/pdns/0.0.7)

**Comment by pieterlexis** [comment]
Hi @ffledgling ,

I had to apply the following patch to the specification to make `swagger-codegen` actually generate code:

```diff
--- authoritative-api-swagger.yaml	2017-11-15 14:31:29.792356427 +0100
+++ authoritative-api-swagger-fixed.yaml	2017-11-15 14:30:34.813307561 +0100
@@ -447,7 +447,7 @@
           schema:
             type: array
             items:
-              - $ref: '#/definitions/Metadata'
+              $ref: '#/definitions/Metadata'
     post:
       summary: 'Creates a set of metadata entries'
       description: 'Creates a set of metadata entries of given kind for the zone. Existing metadata entries for the zone with the same kind are not overwritten.'
@@ -471,7 +471,7 @@
           schema:
             type: array
             items:
-              - $ref: '#/definitions/Metadata'
+              $ref: '#/definitions/Metadata'
       responses:
         '204':
           description: OK
@@ -504,7 +504,7 @@
           schema:
             type: array
             items:
-              - $ref: '#/definitions/Metadata'
+              $ref: '#/definitions/Metadata'
     put:
       summary: 'Modify the content of a single kind of domain metadata.'
       operationId: modifyZoneMetadataKind
@@ -527,7 +527,7 @@
           schema:
             type: array
             items:
-              - $ref: '#/definitions/Metadata'
+              $ref: '#/definitions/Metadata'
       responses:
         '204':
           description: OK
@@ -579,7 +579,7 @@
           schema:
             type: array
             items:
-              - $ref: '#/definitions/Cryptokey'
+              $ref: '#/definitions/Cryptokey'
     post:
       summary: 'Creates a Cryptokey'
       description: 'This method adds a new key to a zone. The key can either be generated or imported by supplying the content parameter. if content, bits and algo are null, a key will be generated based on the default-ksk-algorithm and default-ksk-size settings for a KSK and the default-zsk-algorithm and default-zsk-size options for a ZSK.'
@@ -603,7 +603,7 @@
           schema:
             type: array
             items:
-              - $ref: '#/definitions/Cryptokey'
+              $ref: '#/definitions/Cryptokey'
       responses:
         '204':
           description: OK
@@ -636,7 +636,7 @@
           schema:
             type: array
             items:
-              - $ref: '#/definitions/Cryptokey'
+              $ref: '#/definitions/Cryptokey'
     put:
       summary: 'This method (de)activates a key from zone_name specified by cryptokey_id'
       operationId: modifyCryptokey
@@ -659,7 +659,7 @@
           schema:
             type: array
             items:
-              - $ref: '#/definitions/Cryptokey'
+              $ref: '#/definitions/Cryptokey'
       responses:
         '204':
           description: OK
@@ -926,9 +926,9 @@
 # FIXME: This is problematic at the moment, because swagger doesn't support this type of mixed response
 #  SearchResult:
 #    anyOf:
-#      - $ref: '#/definitions/SearchResultZone'
-#      - $ref: '#/definitions/SearchResultRecord'
-#      - $ref: '#/definitions/SearchResultComment'
+#      $ref: '#/definitions/SearchResultZone'
+#      $ref: '#/definitions/SearchResultRecord'
+#      $ref: '#/definitions/SearchResultComment'
 
 # Since we can't do 'anyOf' at the moment, we create a 'superset object'
   SearchResult:
@@ -954,7 +954,7 @@
   SearchResults:
     type: array
     items:
-      - $ref: '#/definitions/SearchResult'
+      $ref: '#/definitions/SearchResult'
 
   Metadata:
     properties:
```

**Comment by pieterlexis** [comment]
Note, I updated the spec so the cryptokey and metadata endpoints function properly, file is [here](https://dump.plexis.eu/swagger.yaml)

**Comment by pieterlexis** [comment]
I actually discovered and inconsistency with the PowerDNS API when working with the swagger file. This is fixed in #5964 . The following diff works correctly after that PR is merged:

```diff
--- swagger.yaml	2017-11-16 16:17:25.993236135 +0100
+++ swagger-0.0.10.yaml	2017-11-16 16:17:53.439449735 +0100
@@ -1,6 +1,6 @@
 swagger: '2.0'
 info:
-  version: "0.0.9"
+  version: "0.0.10"
   title: PowerDNS Authoritative HTTP API
   license:
     name: MIT
@@ -987,9 +987,14 @@
       ds:
         type: array
         items:
-          # 'TODO: What is a DS record?'
-          type: object
+          type: string
         description: 'An array of DS records for this key'
       privatekey:
         type: string
-        description: 'The private key in ISC format'
\ No newline at end of file
+        description: 'The private key in ISC format'
+      algorithm:
+        type: string
+        description: 'The name of the algorithm of the key, should be a mnemonic'
+      bits:
+        type: integer
+        description: 'The size of the key'
```

**Comment by ffledgling** [comment]
Updated https://app.swaggerhub.com/apis/ffledgling/pdns/0.0.10

**Comment by pieterlexis** [comment]
I added a title and a description to the objects, so it will be easier to generate docs from the spec:

```diff
--- ../pdns/docs/http-api/swagger/authoritative-api-swagger.yaml	2017-11-21 11:46:26.874302604 +0100
+++ swagger.yaml	2017-11-21 16:35:33.635352808 +0100
@@ -1,6 +1,6 @@
 swagger: '2.0'
 info:
-  version: "0.0.12"
+  version: "0.0.13"
   title: PowerDNS Authoritative HTTP API
   license:
     name: MIT
@@ -696,6 +696,7 @@
 
 definitions:
   Server:
+    title: Server
     properties:
       type:
         type: string
@@ -725,6 +726,8 @@
       $ref: '#/definitions/Server'
 
   Zone:
+    title: Zone
+    description: This represents an authoritative DNS Zone.
     properties:
       id:
         type: string
@@ -800,6 +803,8 @@
       $ref: '#/definitions/Zone'
 
   RRSet:
+    title: RRSet
+    description: This represents a Resource Record Set (all records with the same name and type).
     required:
       - name
       - type
@@ -831,6 +836,8 @@
           $ref: '#/definitions/Comment'
 
   Record:
+    title: Record
+    description: The RREntry object represents a single record.
     required:
       - content
       - disabled # PatchZone endpoint complains if this is missing
@@ -846,6 +853,8 @@
         description: 'If set to true, the server will find the matching reverse zone and create a PTR there. Existing PTR records are replaced. If no matching reverse Zone, an error is thrown. Only valid in client bodies, only valid for A and AAAA types. Not returned by the server.'
 
   Comment:
+    title: Comment
+    description: A comment about an RRSet.
     properties:
       content:
         type: string
@@ -858,6 +867,7 @@
         description: 'Timestamp of the last change to the comment'
 
   ConfigSetting:
+    title: ConfigSetting
     properties:
       name:
         type: string
@@ -870,6 +880,7 @@
         description: 'The value of setting name'
 
   StatisticItem:
+    title: StatisticItem
     properties:
       name:
         type: string
@@ -882,6 +893,7 @@
         description: 'The value of item'
 
   SearchResultZone:
+    title: SearchResultZone
     properties:
       name:
         type: string
@@ -892,6 +904,7 @@
         type: string
 
   SearchResultRecord:
+    title: SearchResultRecord
     properties:
       content:
         type: string
@@ -912,6 +925,7 @@
         type: integer
 
   SearchResultComment:
+    title: SearchResultComment
     properties:
       content:
         type: string
@@ -934,6 +948,7 @@
 
 # Since we can't do 'anyOf' at the moment, we create a 'superset object'
   SearchResult:
+    title: SearchResult
     properties:
       content:
         type: string
@@ -959,6 +974,8 @@
       $ref: '#/definitions/SearchResult'
 
   Metadata:
+    title: Metadata
+    description: Represents zone metadata
     properties:
       kind:
         type: string
@@ -970,6 +987,8 @@
         description: 'Array with all values for this metadata kind.'
 
   Cryptokey:
+    title: Cryptokey
+    description: 'Describes a DNSSEC cryptographic key'
     properties:
       type:
         type: string
```

File: docs/http-api/swagger/authoritative-api-swagger.yaml [hunk 11]
```diff
@@ -0,0 +1,1002 @@
...
+      responses:
+        '201':
+          description: Created
+          schema:
+            $ref: '#/definitions/Cryptokey'
+
+  '/servers/{server_id}/zones/{zone_id}/cryptokeys/{cryptokey_id}':
+    get:
+      summary: 'Returns all data about the CryptoKey, including the privatekey.'
+      operationId: listCryptokey
```
Comment (pieterlexis) [inline docs/http-api/swagger/authoritative-api-swagger.yaml]: should be renamed to getCryptokey

**Pull request merged**